### PR TITLE
bluetooth: kw41: Add KW41Z BLE support

### PR DESCRIPTION
**Pull request by dleach02**
- Add build changes to support NXP BLE link layer library to Zephyr
- Add BLE link layer controller wrapper around NXP BLE library

Signed-off-by: David Leach <david.leach@nxp.com>

### CODE REVIEW

**Comment by carlescufi** [comment]
@dleach02 there seems to be some issues with your commit summaries. We typically add spaces after the colons (`ext: mcux:` and not `ext:mcux`) and we also use `Bluetooth: controller` and `Bluetooth: host` and not `ble:`.

File: ext/hal/nxp/mcux/middleware/wireless/framework_5.3.3/OSAbstraction/Interface/fsl_os_abstraction.h
Comment (carlescufi) [inline ext/hal/nxp/mcux/middleware/wireless/framework_5.3.3/OSAbstraction/Interface/fsl_os_abstraction.h]: We don't typically change files in `ext/` directly, instead importing a new revision from upstream.

File: ext/hal/nxp/mcux/middleware/wireless/framework_5.3.3/OSAbstraction/Source/fsl_os_abstraction_zephyr.c
Comment (carlescufi) [inline ext/hal/nxp/mcux/middleware/wireless/framework_5.3.3/OSAbstraction/Source/fsl_os_abstraction_zephyr.c]: Couldn't you upstream this file?

File: ext/hal/nxp/mcux/middleware/wireless/framework_5.3.3/OSAbstraction/Interface/fsl_os_abstraction.h
Comment (dleach02) [reply ext/hal/nxp/mcux/middleware/wireless/framework_5.3.3/OSAbstraction/Interface/fsl_os_abstraction.h]: Understood. Part of the problem is the initial version of this file only included the minimum that the XCVR driver needed (before my time) so I continued adding the functions needed by the BLE link layer library. My preference right now is to not include all of the abstraction as we will likely not be pulling in any additional code from NXP that will need it.

File: boards/arm/frdm_kw41z/frdm_kw41z.dts
Comment (jfischer-phytec-iot) [inline boards/arm/frdm_kw41z/frdm_kw41z.dts]: while you are at it, can you please add a boot_partition?

File: boards/arm/frdm_kw41z/frdm_kw41z.dts
Comment (jfischer-phytec-iot) [inline boards/arm/frdm_kw41z/frdm_kw41z.dts]: Please fix it in previous commit.

File: boards/arm/frdm_kw41z/frdm_kw41z.dts
Comment (dleach02) [reply boards/arm/frdm_kw41z/frdm_kw41z.dts]: I'm not sure how to test something like that? How big would I make it? 

The KW41Z never seemed to be built with a boot partition but the BLE stuff needed a file storage partition so I just added what was needed for the BLE stuff.

File: subsys/bluetooth/controller/Kconfig
Comment (joerchan) [inline subsys/bluetooth/controller/Kconfig]: Organization-wise I think it would be better to separate all of these options into a separate Kconfig file.
controller/hci/vendor/<nxp_vendor_name>/Kconfig would be my preference.

File: subsys/bluetooth/controller/hci/vendor/hci_kw41z.c
Comment (joerchan) [inline subsys/bluetooth/controller/hci/vendor/hci_kw41z.c]: Since there is more than one file in vendor, I think it would be better to have a subfolder for each vendor, to get a better organization.
So hci/vendor/<nxp_vendor_name>/<sources>
If that would make the subfolder nesting too deep, I think it would make it better to drop the vendor folder and just have: hci/<vendor1>, hci/<vendor2>

File: ext/hal/nxp/mcux/CMakeLists.txt
Comment (dleach02) [reply ext/hal/nxp/mcux/CMakeLists.txt]: Good point. Fixed

File: subsys/bluetooth/controller/Kconfig
Comment (dleach02) [reply subsys/bluetooth/controller/Kconfig]: yes, I agree.

File: subsys/bluetooth/controller/CMakeLists.txt
Comment (dleach02) [reply subsys/bluetooth/controller/CMakeLists.txt]: That looks correct but this was how it was structured before so I didn't change any of this part. i.e., not sure why it was like this in the first place?

File: subsys/bluetooth/controller/hci/vendor/hci_kw41z.c
Comment (Vudentz) [reply subsys/bluetooth/controller/hci/vendor/hci_kw41z.c]: Id drop the vendor portion and just have the actual vendor name instead.

Comment (dleach02) [reply subsys/bluetooth/controller/hci/vendor/hci_kw41z.c]: Agreed

File: subsys/bluetooth/controller/CMakeLists.txt
Comment (cvinayak) [reply subsys/bluetooth/controller/CMakeLists.txt]: You could make it similar to left hand side, as it is in the upstream master.

**Review by cvinayak — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/common/Kconfig
Comment (jhedberg) [inline subsys/bluetooth/common/Kconfig]: This will conflict with #10264. Any chance you could just remove it from the PR?

Comment (jhedberg) [reply subsys/bluetooth/common/Kconfig]: Or to put it another way: at least follow the convention of existing HCI drivers and logging instead of introducing this piece that will anyway be removed once #10264 gets merged.

Comment (dleach02) [reply subsys/bluetooth/common/Kconfig]: I marked this PR as DNM because I was aware of a couple of different activities that would impact various things including the logging changes. At the time of this edit, there wasn't a way to just enable logging for the HCI driver since it appeared to be tied to the BLE logging so I added this. There are some internal NXP things we are trying to work out for hosting the library which will likely take longer then the #10264 PR so, yes, I will move this to the new system when it is ready.

File: subsys/bluetooth/controller/CMakeLists.txt
Comment (dleach02) [reply subsys/bluetooth/controller/CMakeLists.txt]: I'm not following what you are requesting? I have to wrap these sources with the BT_LL_SW test because these files are not needed when building for the KW41 link layer library.

File: subsys/bluetooth/common/Kconfig
Comment (dleach02) [reply subsys/bluetooth/common/Kconfig]: By the way, I did start making changes to align with the new zephyr logging system going in but stopped to wait instead on when the BT subsystem catches up.

File: subsys/bluetooth/controller/CMakeLists.txt
Comment (cvinayak) [reply subsys/bluetooth/controller/CMakeLists.txt]: in line 4 there is zephyr_library_sources, and you add it again in line 16. On the left hand side, in the parent commit, the ctrl.c and ll.c is part of the single zephyr_library_sources. 

also, you have removed CONFIG_BT_CTLR_CRYPTO dependency for the crypto.c file.

**Review by cvinayak — CHANGES_REQUESTED** [review]

File: boards/arm/frdm_kw41z/doc/frdm_kw41z.rst
Comment (dbkinder) [inline boards/arm/frdm_kw41z/doc/frdm_kw41z.rst]:
```suggestion
with the NXP-provided BLE Link Layer Controller library. The library will need to
```

File: boards/arm/frdm_kw41z/doc/frdm_kw41z.rst
Comment (dbkinder) [inline boards/arm/frdm_kw41z/doc/frdm_kw41z.rst]: better would be:

    be manually copied from the KW41Z SDK to this directory:
    ``<zephyr base>/ext/lib/ble/nxp_kinetis``

File: boards/arm/frdm_kw41z/doc/frdm_kw41z.rst
Comment (dbkinder) [inline boards/arm/frdm_kw41z/doc/frdm_kw41z.rst]:
```suggestion
``FRDM-KW41Z.zip`` SDK file. Copy the following library file from the extracted zip
```

File: boards/arm/frdm_kw41z/doc/frdm_kw41z.rst
Comment (dbkinder) [inline boards/arm/frdm_kw41z/doc/frdm_kw41z.rst]: better would be:

    file into the directory referenced above:
    ``<FRDM-KW41Z.zip>\middleware\wireless\bluetooth_1.2.6\controller\lib\lib_ble_kw41z_controller.a``

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: boards/arm/frdm_kw41z/doc/frdm_kw41z.rst
Comment (dleach02) [reply boards/arm/frdm_kw41z/doc/frdm_kw41z.rst]: I don't believe that is the proper use of a hyphen.

**Comment by aescolar** [comment]
@dleach02 : Is there interest in continuing with this PR? Otherwise would you like to close it?

**Comment by dleach02** [comment]
@aescolar there are some third party folks that have been playing with this PR. Internally at NXP, we are still working through the logistics on how to distribute a library used by this PR. For now, I would prefer to just leave the PR as is and as DNM until NXP works out the internal logistics.

**Comment by aescolar** [comment]
@dleach02 Would it be acceptable for you to close this PR? (It can always be reopened, and its base branch would remain in your own clone. Closing it simply reduces the amount of noise maintainers need to sift thru)

**Comment by dleach02** [comment]
Closing this PR until internal logistics are worked out in NXP on how to handle library deployment.

**Pull request closed without merging**
